packages/react-storage/src/components/StorageBrowser/controls/hooks/useDataTable.ts

```diff
@@ -0,0 +1,170 @@
+import React from 'react';
+import {
+  DataTableButtonDataCell,
+  DataTableDateDataCell,
+  DataTableTextDataCell,
+  DataTableDataCell,
+  DataTableProps,
+  SortDirection,
+  DataTableNumberDataCell,
+} from '../../composables/DataTable';
+import { useControlsContext } from '../context';
+import { compareButtonData } from '../compareFunctions/compareButtonData';
+import { compareDateData } from '../compareFunctions/compareDateData';
+import { compareNumberData } from '../compareFunctions/compareNumberData';
+import { compareTextData } from '../compareFunctions/compareTextData';
+
+export type UseDataTable = () => {
+  props?: DataTableProps;
+};
+
+interface SortState {
+  index: number;
+  direction: SortDirection;
+}
+
+type GroupedRows = {
+  [K in DataTableDataCell['type']]: DataTableProps['rows'];
+};
+
+const GROUP_ORDER: DataTableDataCell['type'][] = [
```

Comment:
It is possible for rows to contain different cell types at the same index. To accommodate this in a reasonable manner, we first group the rows by the cell types under sort before sorting the groups individually. The group order then determines the priority in which the sorted groups are displayed.